File: processor/src/main/java/io/shreyash/rush/blocks/Method.java
package io.shreyash.rush.blocks;

import com.google.appinventor.components.annotations.SimpleFunction;

import javax.annotation.processing.Messager;
import javax.lang.model.element.Element;
import javax.lang.model.element.ExecutableElement;
import javax.tools.Diagnostic;

import io.shreyash.rush.util.Casing;
import shaded.org.json.JSONArray;
import shaded.org.json.JSONObject;

public class Method extends BlockWithParams {
  private final ExecutableElement element;
  private final Messager messager;

  public Method(Element element, Messager messager) {
    super(element);
    this.element = (ExecutableElement) element;
    this.messager = messager;
    validate();
  }

  @Override
  String description() {
    return this.element.getAnnotation(SimpleFunction.class).description();
  }

  @Override
  void validate() {
    // Check method name
    if (!Casing.isPascalCase(this.name())) {
      messager.printMessage(
          Diagnostic.Kind.WARNING,
          "Simple function \"" + this.name() + "\" should follow 'PascalCase' naming convention."
      );
    }

    // Check param names
    this.params().forEach(el -> {
      if (!Casing.isCamelCase(el.getName())) {
        messager.printMessage(
            Diagnostic.Kind.WARNING,
            "Parameter \"" + el.getName() + "\" in simple function \"" + this.name() + "\" should " +
            "follow 'camelCase' naming convention."
        );
      }
    });
  }

  /**
   * JSON structure:
   * {
   *   "name": "Foo",
   *   "description": "This is a description",
   *   "deprecated": "false",
   *   "returnType": "any",
   *   "params": [
   *      { "name": "bar", "type": "number" },
   *   ]
   *  }
   */
  @Override
  public JSONObject asJsonObject() {
    final JSONObject methodJson = new JSONObject();

    methodJson.put("name", this.name());
    methodJson.put("description", this.description());
    methodJson.put("deprecated", Boolean.toString(this.element.getAnnotation(Deprecated.class) != null));

    // Here, null represents the return type is void. Return type for void
    // methods don't need to be specified
    if (this.returnType() != null) {
      methodJson.put("returnType", this.returnType());
    }

    final JSONArray params = new JSONArray();
    this.params().forEach(el -> {
      final JSONObject paramObj = new JSONObject();
      paramObj.put("name", el.getName());
      paramObj.put("type", el.getType());

      params.put(paramObj);
    });

    methodJson.put("params", params);
    return methodJson;
  }
}

File: processor/src/main/java/io/shreyash/rush/block/Method.kt
package io.shreyash.rush.block

import com.google.appinventor.components.annotations.SimpleFunction
import javax.annotation.processing.Messager
import javax.lang.model.element.Element
import javax.lang.model.element.ExecutableElement
import javax.tools.Diagnostic
import io.shreyash.rush.util.isCamelCase
import io.shreyash.rush.util.isPascalCase
import shaded.org.json.JSONObject

class Method(element: Element, private val messager: Messager) : BlockWithParams(element) {
    private val element = element as ExecutableElement

    init {
        runChecks()
    }

    override fun description() = this.element.getAnnotation(SimpleFunction::class.java).description

    override fun runChecks() {
        // Check method name
        if (!isPascalCase(name())) {
            messager.printMessage(
                Diagnostic.Kind.WARNING,
                "Simple function \"" + name() + "\" should follow 'PascalCase' naming convention."
            )
        }

        // Check param names
        params().forEach {
            if (!isCamelCase(it.name)) {
                messager.printMessage(
                    Diagnostic.Kind.WARNING,
                    "Parameter \"" + it.name + "\" in simple function \"" + name() + "\" should " +
                            "follow 'camelCase' naming convention."
                )
            }
        }
    }

    /**
     * @return JSON representation of this method.
     * {
     *  "name": "Foo",
     *  "description": "This is a description",
     *  "deprecated": "false",
     *  "returnType": "any",
     *  "params": [
     *    { "name": "bar", "type": "number" },
     *  ]
     * }
     */
    override fun asJsonObject(): JSONObject {
        val methodJson = JSONObject()
            .put("name", name())
            .put("description", description())
            .put("deprecated", this.deprecated().toString())

        // Here, null represents the return type is void. Return type for void methods don't need to
        // be specified
        if (returnType() != null) {
            methodJson.put("returnType", returnType())
        }

        val params = params().map {
            JSONObject()
                .put("name", it.name)
                .put("type", it.type)
        }
        methodJson.put("params", params)

        return methodJson
    }
}
